typing/typedecl.ml

```diff
@@ -1462,10 +1462,37 @@ let get_native_repr_attribute attrs ~global_repr =
   | _, Some { Location.loc }, _ ->
     raise (Error (loc, Multiple_native_repr_attributes))
 
-let native_repr_of_type env kind ty =
+let is_constants_only_variant path env =
+  let open Types in
+  let decl = Env.find_type path env in
+  let constants cd = match cd.cd_args with
+    | Cstr_tuple [] -> true
+    | _ -> false
+  in
+  match decl.type_kind with
+  | Type_abstract | Type_record _ | Type_open -> false
+  | Type_variant(crts, _) -> List.for_all constants crts
+
+let is_bigarray_c_layout ty env =
+  let (_, layout) = Typeopt.bigarray_type_kind_and_layout env ty in
+  layout = Lambda.Pbigarray_c_layout
+
+let native_repr_of_type env kind ty ~result =
   match kind, get_desc (Ctype.expand_head_opt env ty) with
   | Untagged, Tconstr (path, _, _) when Path.same path Predef.path_int ->
-    Some Untagged_int
+    Some (Untagged_int Int)
+  | Untagged, Tconstr (path, _, _) when Path.same path Predef.path_bool ->
+    Some (Untagged_int Bool)
+  | Untagged, Tconstr (path, _, _) when Path.same path Predef.path_char ->
+    Some (Untagged_int Char)
+  | Untagged, Tconstr (path, _, _) when is_constants_only_variant path env ->
+    Some (Untagged_int Constants)
```

Comment:
We already have logic in Typeopt to tell whether a type is an immediate or not (one can use `Typeopt.maybe_pointer_type ty`, or `Typeopt.classify` but that one should be exposed in the .mli). One should be able to request untagging for all immediate types. Do you have an idea of a better, simpler implementation that reuses this existing check? Is it actually important to remember which exact immediate type we are untagging, or could all this be handled as a single `Untagged_immediate` constructor for `native_repr`?

Reply:
Good idea. But there is an equality test on the type, probably used to compare external declaration in ml and mli ?
I think keeping the path could work, checking with the appropriate env.